### ISSUE TITLE
Null safety

### ISSUE THREAD

**Issue by Francessco121** [Francessco121/msal-js-dart#18]
Package maintainers are now encouraged to migrate packages to use null safety even though the feature is in beta.
Migration guide: https://dart.dev/null-safety/migration-guide

 Checklist:
- [X] Wait for dependencies to publish a null safe version
  - [X] `js`
  - [X] `pedantic`
  - [X] `test`
- [X] Decide on version number. Guide says it should be considered a breaking change, bringing this package to 2.0.0... this will however take it out of line with msal.js... 😕
  - Applications should be using an exact version number anyway to make sure it matches their install of msal.js, releasing it with a normal version (ex. 1.4.0) may not cause problems.
  - **Resolution:** Null safety will be available as version `1.4.0-nullsafety` and stay as a prerelease version only for MSAL 1.x. When this package moves on to support MSAL 2.x, null safety will be supported by default.
- [x] Migrate (#21)
- [x] Test

**Comment by nyarian** [comment]
kinda have it [here](https://github.com/Three-Horsemen-of-Software-Engineering/msal-js-dart/tree/infrastructure/null-safety-migration), but review would be appreciated
@Francessco121 if you would have time to pinpoint the changes necessary then I'll apply 'em. all unit-tests are currently passing

**Comment by Francessco121** [comment]
@nyarian Really appreciate the contribution! Could you open a pull request to make this easier to review?

The biggest issue I see right now with your changes is that almost everything is now marked as nullable. There's definitely a good amount of fields from msal.js that are never null/undefined. This package should limit the number of nullable values exposed publicly for ease-of-use where possible. Ideally, `package:js` would provide runtime null checking for anything marked as non-nullable on the Dart side but I'm not sure how that behaves yet. 

Unfortunately, this really comes down to either keeping everything nullable and letting the user check it (be quite painful to use the library, but safe), or look very carefully at msal.js and determine which methods/properties are guaranteed to never be null and reflect that.

There's also a couple spots where we could make better use of null safety, such as:
```dart
factory UserAgentApplication(Configuration? configuration) {
    if (configuration == null) throw ArgumentError.notNull('configuration');
// ...
``` 
could just be:
```dart
factory UserAgentApplication(Configuration configuration) {
// ...
```

Lastly is still the concern about the version number. Given that MSAL 2.x would be nice to support (#19) and keeping this package's version inline with MSAL is ideal, it may be better to mark this release as `1.4.0-nullsafety.0`. It would break semantic versioning however since this is a breaking change. Keeping MSAL 1.x null safety as a prerelease version only may help?

**Comment by nyarian** [comment]
stating that this migration is naive would be an understatement :)
I'll do a proper refactoring pass this evening and will open the MR after. the thing that keeps me behind here is that I'm not much in TS and web in general, so it's a bit tricky to define which things are non-nullable by design in the original `msal-js` library.
concerning versioning - yeah, that's reasonable, I tend to agree. I believe it should be as close as possible to the upstream `msal-js` versions. I would suggest to just postfix the version that's currently mirrored by the library with `-nullsafety.0`, then, so that it's `1.3.2-nullsafety.0`. what do you think?

**Comment by Francessco121** [comment]
Current version is msal.js 1.4.5, so given that msal.js 1.4 didn't add anything and is compatible with 1.3 of this package, I think jumping to `1.4.0-nullsafety.0` makes sense.

**Comment by nyarian** [comment]
yessir

**Comment by Francessco121** [comment]
Null safety is (finally) tested and published as version `1.4.0-nullsafety.0`!
